### ISSUE TITLE
Will this cause problems for your Trakt API calls?

### ISSUE THREAD

**Issue by eengert** [jurialmunkey/plugin.video.themoviedb.helper#828]
Are you aware of this:  [Reddit Post](https://www.reddit.com/r/Addons4Kodi/comments/ww7tvr/no_limits_plugin_service_tmdb_helper_trakt_limits/?utm_source=share&utm_medium=ios_app&utm_name=iossmf) ?

Im wondering if this could threaten TMDBH Trakt usage.  Just wanted to make sure you knew about it in case it’s a problem.

**Comment by jurialmunkey** [comment]
Yeah I know about this guy but first I've seen of this new service plugin he's written. Unfortunately not much I can do about it. Shame people are selfish. 

The most concerning part is the security risk of people installing a service which injects code into another plugin.

Pretty disappointing. Really kills my enthusiasm when people do things like this.

**Comment by eengert** [comment]
Do you think he understands the potential ramifications of what he's doing?

**Comment by jurialmunkey** [comment]
Definitely not. The script is a really crude find and replace. If I use the variable name "limit" for something else his script is going to break things

**Comment by jurialmunkey** [comment]
Thanks for letting me know btw.

Closed this since there's not much I can do about it. Overall I'm not too concerned other than the fact the it is a obviously a bad idea for a user to install this script considering what it does.

In terms of API ramifications, TMDb will just rate limit the user far more frequently. At the extreme end of the settings, the TMDb server will see thousands of simultaneous requests from the user and its DDOS protection will kick in to block the IP address. 

Obviously not great if every single TMDbH user did that (probably TMDb would disable my key temporarily in that case and get in touch to find out what was going on). But I would think most users are sensible and realise why attempting to retrieve many hundreds of items all at once is a bad idea, so probably wouldn't get to that stage (I hope!)

Trakt will be fine as it is only one request to retrieve the base list for these calls - so no changes on that front. The point of the limit is to reduce the number of simultaneous calls being made to TMDb since each item needs an individual lookup (and making hundreds of calls like that all at once will make triggering those server protections more likely and frequent)

**Comment by ambulancePilot** [comment]
The question I have is how other add-ons don't limit to 20 and don't result in errors. I use one of them, and never receive any errors regarding API limits. I load about 20 widgets each limited to 200 items several times a day and never have an API issue. On top of that, the widgets display in under 10 seconds.

Is it possible that the widget is loading to the point where it can be displayed in Kodi but that no details are retrieved until the user scrolls to the item?

I honestly used to load lists containing upwards of 500 items and still have never had an issue.

I am not criticizing, nor am I asking for you to include a new feature. I am just trying to understand how this all works. I have always been under the understanding that if you load a list of 40, it will take twice as long as loading a list of 20, but so many add-ons have widgets load in under 10 seconds and it does not matter how many items the list has. How?

**Comment by jurialmunkey** [comment]
The main issue is not about loading time (at least since tmdb removed the need for client side limiting). Threading of multiple requests significantly cuts down on loading *as long as the server can cope with the increase of requests*. The main issue is about respecting services that provide this data for free and avoiding unnecessary strain. You loading 200 items per list means that you take up the server resources of 10 normal users. If everyone did what you're doing these services would go down, get slower,  get shut off or put behind a pay wall because the servers cannot deal with the increased strain and the increased cost is too high. 

Trakt went down several times for a significant period last year due to addons allowing this type of behaviour. They also disabled many parts of their API like artwork in 2016 for the same reason. Next year they are putting more features behind pay walls, and will be limiting custom list sizes for non-VIP users. Fanarttv frequently goes down because of server load. Tvdb switched to a paid subscription model. 

We will continue to lose more services that provide free data because people are selfish and don't respect them. TMDbHelper has over 100k daily unique users so a 10x increase in traffic by increasing list sizes is a massive additional strain. 

There is absolutely no need to view 200 items at once. Typically users don't scroll far horizontally to find content - so all that extra loaded data is wasted. Having several smaller curated lists and tailored recommendations is far more important.

And to answer your question - no it is not possible to load data on a rolling basis. Kodi's plugin system simply does not have support for such a feature. Kodi requires metadata for all loaded items before it displays the list (artwork is the only part that is different and loaded on an as needed basis).

**Comment by ambulancePilot** [comment]
Thanks for taking the time to answer. I appreciate it.

**Comment by jurialmunkey** [comment]
Also note that there are already options in TMDbH settings to increase the number of items for Trakt collection and other user specific trakt lists (eg inprogress, watched history, watchlist etc. ) . 

It is only generic lists (eg popular/trending) and custom lists that are limited to 20 item pages 

The thinking behind allowing a larger page limit for user based sync lists is because the majority of the items in these lists don't change and will be cached already on subsequent lookups eg watched history only needs to retrieve all the items once, with additional lookups only needed for newly watched items - so the overall load is dispersed.

**Comment by ambulancePilot** [comment]
Thanks again for clarifying. The part I don't understand is why custom
lists are limited to 20. These lists don't change either, and these are the
lists I am referring to.

I completely understand the need to limit dynamic lists that change
frequently such as trending and popular, but if I have a list of a few
hundred items and that list is essentially static, I just can't imagine the
need to enforce a 20 item limit on something like that. The list should
basically load once and then be cached for a week, no?

Perhaps I am misunderstanding how users on Trakt utilize custom lists. Mine
are almost entirely static, perhaps a new title being added once a week or
two.

**Comment by jurialmunkey** [comment]
The problem is that some custom lists have many thousands of items. For instance, this custom list has 6000+ items. Put that on random sort and you are getting new items every time it loads.
https://trakt.tv/users/theultimatec0der/lists/all-movies-2019?sort=rank,asc

The other issue is that generally custom lists often don't share much content. For instance, a list of Kids Movies and a list of Horror movies aren't going to overlap much. By contrast a list of Recently Watched Shows and a list of In-Progress Shows will have a lot of overlap which cuts down on the number of requests being made in total if you have multiple widgets.

And really it is multiple widget setups where large counts are a problem. One list at a time it isn't so bad to do 100 items. 10+ lists of 100 items is considerably worse.
